### PR TITLE
[ADDED] Support for multiple subject filters on consumers

### PR DESCRIPTION
**Pull request by piotrpio**
Addresses https://github.com/nats-io/nats-architecture-and-design/issues/196

### CODE REVIEW

File: jsm.go [hunk 1]
```diff
@@ -408,6 +412,11 @@ func (js *js) upsertConsumer(stream, consumerName string, cfg *ConsumerConfig, o
 		}
 		return nil, info.Error
 	}
+
+	// check whether multiple filter subjects (if used) are reflected in the returned ConsumerInfo
+	if len(cfg.FilterSubjects) != 0 && len(info.Config.FilterSubjects) == 0 {
```
Comment (wallyqs) [inline jsm.go:417]: To test this I guess would need to run against a version of the server that does not support filter subjects 🤔

File: jsm.go [hunk 2]
```diff
@@ -408,6 +412,11 @@ func (js *js) upsertConsumer(stream, consumerName string, cfg *ConsumerConfig, o
...
+		return nil, ErrConsumerMultipleFilterSubjectsNotSupported
```
Comment (wallyqs) [inline jsm.go:418]: At this point I think it would have have created the consumer but instead returning an error? Wondering how we should we handle creating durable consumers unintentionally that are not used, or document in the error that may have to cleanup the consumer that got created (or return both consumer info and error?).

File: test/js_test.go [hunk 3]
```diff
@@ -5224,6 +5250,50 @@ func TestJetStreamSubscribe_RateLimit(t *testing.T) {
...
+		js.Publish("foo", []byte("msg"))
+	}
+	for i := 0; i < 5; i++ {
+		js.Publish("bar", []byte("msg"))
+	}
+	for i := 0; i < 5; i++ {
+		js.Publish("baz", []byte("msg"))
+	}
+
+	sub, err := js.SubscribeSync("", nats.BindStream("TEST"), nats.ConsumerFilterSubjects("foo", "baz"))
```
Comment (wallyqs) [inline test/js_test.go:5280]: is this an ephemeral consumer? do we need another one with a durable name to cover some of the conditions above?

File: jsm.go (hunk 1 above)
Comment (piotrpio) [reply jsm.go:417]: Yup, this is basically server compatibility check.

**Review by Jarema — CHANGES_REQUESTED** [review]
Looks good, with some small nits.

File: jsm.go [hunk 4]
```diff
@@ -375,9 +375,13 @@ func (js *js) upsertConsumer(stream, consumerName string, cfg *ConsumerConfig, o
...
-	} else if !js.nc.serverMinVersion(2, 9, 0) || (cfg.Durable != "" && js.opts.featureFlags.useDurableConsumerCreate) {
+	} else if !js.nc.serverMinVersion(2, 9, 0) ||
+		(cfg.Durable != "" && js.opts.featureFlags.useDurableConsumerCreate) {
 		// if server version is lower than 2.9.0 or user set the useDurableConsumerCreate flag, use the legacy DURABLE.CREATE endpoint
 		ccSubj = fmt.Sprintf(apiDurableCreateT, stream, consumerName)
+	} else if cfg.FilterSubjects != nil {
+		// if multiple filter subjects are used, we must use legacy DURABLE.CREATE endpoint
+		ccSubj = fmt.Sprintf(apiDurableCreateT, stream, consumerName)
 	} else {
 		// if above server version 2.9.0, use the endpoints with consumer name
```
[the remark this answers is not in the document]
Comment (piotrpio) [reply jsm.go:386]: The new API for creating consumer was introduced in 2.9.0 from what I know.

[the remark this answers is not in the document]
Comment (Jarema) [reply jsm.go:386]: yes, but so far you need to create multiple filtered consumers with the old API.

File: jsm.go [hunk 5]
```diff
@@ -375,9 +375,13 @@ func (js *js) upsertConsumer(stream, consumerName string, cfg *ConsumerConfig, o
 		ccSubj = fmt.Sprintf(apiLegacyConsumerCreateT, stream)
 	} else if err := checkConsumerName(consumerName); err != nil {
 		return nil, err
-	} else if !js.nc.serverMinVersion(2, 9, 0) || (cfg.Durable != "" && js.opts.featureFlags.useDurableConsumerCreate) {
+	} else if !js.nc.serverMinVersion(2, 9, 0) ||
```
[the remark this answers is not in the document]
Comment (piotrpio) [reply jsm.go:378]: Not sure if I understand. Currently I have:
```go
else if !js.nc.serverMinVersion(2, 9, 0) ||
		(cfg.Durable != "" && js.opts.featureFlags.useDurableConsumerCreate) {
		// if server version is lower than 2.9.0 or user set the useDurableConsumerCreate flag, use the legacy DURABLE.CREATE endpoint
		ccSubj = fmt.Sprintf(apiDurableCreateT, stream, consumerName)
	} else if len(cfg.FilterSubjects) != 0 {
		// if multiple filter subjects are used, we must use legacy DURABLE.CREATE endpoint
		ccSubj = fmt.Sprintf(apiDurableCreateT, stream, consumerName)
```

That means, that if server version is <2.9.0 **or** `len(cfg.FilterSubjects != 0` we will always use legacy durable subject. However, from what I understood, if `serverVersion  >= 2.10.0` and we do not have multiple filters set, we can and should use the new create consumer API. With the change you suggested, we would not do this.

File: jsm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (piotrpio) [reply jsm.go:386]: Not sure if I understand. Currently I have:
```go
else if !js.nc.serverMinVersion(2, 9, 0) ||
		(cfg.Durable != "" && js.opts.featureFlags.useDurableConsumerCreate) {
		// if server version is lower than 2.9.0 or user set the useDurableConsumerCreate flag, use the legacy DURABLE.CREATE endpoint
		ccSubj = fmt.Sprintf(apiDurableCreateT, stream, consumerName)
	} else if len(cfg.FilterSubjects) != 0 {
		// if multiple filter subjects are used, we must use legacy DURABLE.CREATE endpoint
		ccSubj = fmt.Sprintf(apiDurableCreateT, stream, consumerName)
```

That means, that if server version is <2.9.0 **or** `len(cfg.FilterSubjects != 0` we will always use legacy durable subject. However, from what I understood, if `serverVersion  >= 2.10.0` and we do not have multiple filters set, we can and should use the new create consumer API. With the change you suggested, we would not do this.

[the remark this answers is not in the document]
Comment (Jarema) [reply jsm.go:386]: You are right.

File: jsm.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (piotrpio) [reply jsm.go:368]: This check should be for 2.9.0 as it's about verifying if we can use subject with filter subject token when creating consumer (which was 2.9.0)

**Review by Jarema — APPROVED** [review]
LGTM!

**Comment by tpihl** [comment]
@piotrpio I am trying to understand where our new is heading and looking at this pr, It's something I don't get;

We have a new error if filtersubjects is empty? Why?

TLDR; I think it makes more sense to allow it to be empty, and that is an  indication that the consumer should not filter the stream, should get all.



In a little more detail;

If i want to read everything in the stream, do I now have to add '>'. And is that allowed or will that be illegal since it's outside the streams defined subjects.. If it's illegal, do I need to mirrior every subject from the stream definition to get all?

The concept of filtering for me is "no filters means everything". And then filters should possibly be of both types to be able to widen the filter or narrow the filter. To mimic current and my understanding if intended behavior but with the filter-perspective instead of what i would call subset-if-subjects or only-subjects;

empty filtering -> get everything
onlySubjects(subjectwithwildcard,...) -> only msgs with subject matching provided filters
exceptSubjects(subjectswithcildcards...) -> only msgs with subject NOT matching provided filter

Collisions would be avoided by a few simple sorting-criteria so that more explicit filter-parts wins over those with wildcards.

**Comment by piotrpio** [comment]
@tpihl The error will not be returned if you provide empty or nil slice, only if you give an empty string as an element of the slice:
- `FilterSubjects: []string{}` - valid
- `FilterSubjects: nil` - valid
- `FilterSubjects: []string{""}` - invalid

That's a safeguard on the server, and it's treated as invalid configuration. I do not see the point of providing empty string as filter value (in slice), that looks more like unintentional behavior.

> The concept of filtering for me is "no filters means everything".

You're right about this - but there is a difference between providing empty slice (no filters) and providing an empty string as filter.

**Comment by tpihl** [comment]
Okay, was just reading the client code, and didn't understand the error. Maybe something like "filtersubjects cannot contain an empty string" would have been clearer?

/T
